objects/src/assets/mod.rs

```diff
@@ -0,0 +1,420 @@
+
+impl NonFungibleAsset {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a non-fungible asset created from the specified asset details.
+    ///
+    /// # Errors
+    /// Returns an error if the provided faucet ID is not for a non-fungible asset faucet.
+    pub fn new(details: &NonFungibleAssetDetails) -> Result<Self, AssetError> {
+        let asset_hash = Hasher::hash(details.asset_data().as_bytes());
+        Self::from_raw_parts(details.faucet_id(), asset_hash)
+    }
+
+    /// Return a non-fungible asset created from the specified faucet and using the provided
+    /// hash of the asset's data.
+    ///
+    /// Hahs of the asset's data is expected to be computed from the binary representation of the
+    /// asset's data.
+    ///
+    /// # Errors
+    /// Returns an error if the provided faucet ID is not for a non-fungible asset faucet.
+    pub fn from_raw_parts(faucet_id: AccountId, data_hash: Digest) -> Result<Self, AssetError> {
+        if !faucet_id.is_non_fungible_faucet() {
+            return Err(AssetError::not_a_non_fungible_faucet_id(faucet_id));
+        }
+
+        // hash faucet ID and asset data hash together
+        let faucet_id: Word = faucet_id.into();
+        let mut asset: Word = Hasher::merge(&[faucet_id.into(), data_hash]).into();
+
+        // set the first bit of the asset to 0; we can do this because setting the first bit to 0
+        // will always result in a valid field element.
+        asset[0] = Felt::new((asset[0].as_int() << 1) >> 1);
+
+        // set the 32 least significant bits of the 3rd element to the non-fungible asset tag
+        let temp = (asset[2].as_int() >> 32) << 32;
+        asset[2] = Felt::new(temp | AccountId::NON_FUNGIBLE_FAUCET_TAG as u64);
+
+        Ok(Self(asset))
+    }
+
+    // HELPER FUNCTIONS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an error if:
+    /// - The first bit of the asset is not ZERO.
+    /// - The 32 least significant bits of the 3rd element are not set to 2^31.
+    fn validate(&self) -> Result<(), AssetError> {
+        let first_bit = self.0[0].as_int() >> 63;
+        if first_bit != 0 {
+            return Err(AssetError::non_fungible_asset_invalid_first_bit());
+        }
+
+        let tag = self.0[2].as_int() as u32;
+        if tag != AccountId::NON_FUNGIBLE_FAUCET_TAG {
+            return Err(AssetError::NonFungibleAssetInvalidTag(tag));
+        }
+
+        Ok(())
+    }
+}
+
+impl Deref for NonFungibleAsset {
+    type Target = Word;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<NonFungibleAsset> for Word {
+    fn from(asset: NonFungibleAsset) -> Self {
+        asset.0
+    }
+}
+
+impl From<NonFungibleAsset> for [u8; 32] {
+    fn from(asset: NonFungibleAsset) -> Self {
+        let mut result = [0_u8; 32];
```

Comment:
Should we `debug_assert` the first bit to `0`?

Reply:
I'm not sure it is needed here. If we have a `NonFungibleAsset`, I think we can assume that it has been constructed correctly. Otherwise, we should probably assert this every time we use these structs.